### ISSUE TITLE
Improved error handling in code and error return from the API

### ISSUE THREAD

**Issue by t3m8ch** [MergeMinds/mm-backend#20]
На данный момент, если происходит какая-либо ошибка, возвращается пустой ответ. По-хорошему должен возвращаться JSON примерно такого формата.

```JSON
{
  "error": {
    "type": "NOT_FOUND",
    "msg": "User with ID = 5 not found"
  }
}
```

Поле `type` полезно для фронтенда, чтобы было проще писать ветвления для разных видов ошибок. Поле же `msg` содержит дополнительную информацию об ошибке, которая может помочь отладить бекенд.

**Comment by t3m8ch** [comment]
До этого для ошибок использовалась библиотека `thiserror`, которая, как я понял, позволяет не реализовывать вручную трейт `From`. Не знаю, стоит ли в этом PR использовать эту библиотеку или не надо.

Кроме того, мне не нравится, что нужно поголовно везде писать `map_err`, особенно если маппинг происходит к `APIError::InternalServer`. Было бы прикольно, если можно было маппить любую ошибку к `InternalServer` без дополнительной писанины (кроме условного `sqlx::Error`, для которого можно было бы оставить текущую реализацию трейта `From`). Но я вообще не знаю, есть ли такая возможность в Rust.

**Comment by evgenymng** [comment]
> Было бы прикольно, если можно было маппить любую ошибку

Нет, в Rust все типы должны быть явно указаны. Поэтому если ты хочешь все ошибки возвращать из одной функции, у тебя должен быть "наиболее общий тип ошибки".

Разумеется, писать везде явный каст очень неудобно. Поэтому Rust'овый оператор `?` также позволяет приводить ошибку к типу, который возвращает функция, с помощью реализации `From`.

> The `?` operator automatically converts the error to the function's return type error using the `From` trait. This is handy when dealing with functions that return different error types.

https://rustjobs.dev/blog/the-question-mark-operator-in-rust/

**Comment by t3m8ch** [comment]
> > Было бы прикольно, если можно было маппить любую ошибку
> 
> Нет, в Rust все типы должны быть явно указаны. Поэтому если ты хочешь все ошибки возвращать из одной функции, у тебя должен быть "наиболее общий тип ошибки".
> 
> Разумеется, писать везде явный каст очень неудобно. Поэтому Rust'овый оператор `?` также позволяет приводить ошибку к типу, который возвращает функция, с помощью реализации `From`.
> 
> > The `?` operator automatically converts the error to the function's return type error using the `From` trait. This is handy when dealing with functions that return different error types.
> 
> https://rustjobs.dev/blog/the-question-mark-operator-in-rust/

Окей, тогда я просто буду реализовывать `From` для всего, что мне нужно
